### PR TITLE
feat(delivery): bound the composed summary comment with a per-platform byte budget (#324)

### LINKED ISSUE #324: Bound the composed summary comment: the skipped-findings section is uncapped
Surfaced by the #279 pass (PR #323), filed as out-of-scope-but-real per the standing practice.

## Fact

`scripts/post_review.py::compose_review_body` assembles the posted summary comment as brand header, `review_body`, the unanchored-findings section from `build_skipped_section`, then the mechanical footer. After #279 the `review_body` half is bounded (the pipeline renders it from the report Summary, folded at `REPORT_FOLD_LIMITS.summaryChars`), but `build_skipped_section` renders every skipped finding's title, body and suggestion with no size guard, and neither poster checks the composed body against a platform limit (GitHub review bodies cap at 65,536 characters; GitLab notes at 1,000,000).

## Why it matters

A run with many unanchorable findings can push the composed body past the GitHub cap. The API rejects the whole review, so every inline comment is lost along with the footer that carries the incremental-review marker.

## Decision or fix shape

Add a code-owned final-body budget in `post_review.py`: reserve the header and footer bytes first, include only complete skipped entries that fit, append one deterministic omitted-count line, and always keep the mechanical footer last. Pin the exact boundary for both posters and regenerate the bench mirror fixtures that carry a skipped section.

## Verification tier

Suites-only: `tests/test_post_review.py` boundary cases for both posters, bench adapter fixture fidelity.

### PR DESCRIPTION
**Pull request by leehopper**
## Summary

`post_review.py` now bounds the complete summary comment for GitHub reviews and GitLab summary notes with a per-platform UTF-8 byte budget.
A body that fits is byte-identical to today's.
An over-budget body keeps whole consolidation groups in list order, counts what it leaves out, folds an oversized `review_body`, and always ends with the canonical marker footer.
Both posters refuse an over-limit body before the summary POST, and dry-run captures keep their shape.

## Why

`post_review.py` composed the summary without a size check, and the skipped section rendered every unanchorable finding uncapped.
GitHub rejects an overlong review body with 422, so the whole review is lost: every inline comment and the incremental-review marker.
GitLab posts its summary note before any discussion, so a rejected note exits before the discussions post.
On-disk artifacts rendered 1,943 unique findings as skipped entries: p50 641, p99 1,790, and maximum 2,647 characters.
UTF-8 bytes exceeded characters by at most 2.6 percent.
None of 165 dry-run captures carried a skipped section, so the overflow is a reachable worst case rather than an observed one.

## Design

- `PLATFORM_BODY_LIMITS` holds the label, surface and byte limit per platform: GitHub review bodies 65,536 bytes and GitLab summary notes 1,000,000 bytes.
- `compose_review_body(...)` returns `ComposedBody(body, shown, omitted, folded_bytes, omitted_entries)` and owns the skipped section, the footer and the budget.
- The fast path returns today's exact bytes when they fit, including the existing footer deduplication.
- The bounded path reserves the header, separators, the section frame at maximum digit width, the closing line and the full canonical footer before admitting content.
- A consolidation group is the fitting unit, so first-fit in list order shows it whole or omits it whole, and the "listed in full" sentence stays true.
- One closing line reports how many skipped findings are not shown, and it appears only when a group is omitted.
- An oversized `review_body` folds at a line boundary, never splits a code point, drops an unclosed HTML comment, closes an odd code fence and states the dropped bytes.
- When a fold happens, no skipped group is shown, so priority does not flip with summary size.
- The bounded path always appends the full canonical footer, so a fold cannot expose an earlier same-SHA marker with a wrong count.
- `_refuse_over_limit` runs immediately before each summary POST, and GitLab runs it only when it posts a new note.
- Posters print one status line for omissions and one for folds, and warn once per omitted finding on stderr.

## Trade-offs and declined alternatives

GitHub does not document its counting unit, and UTF-8 bytes are never fewer than code points or UTF-16 units.
The byte budget over-omits multi-byte text relative to a code-point limit, and a CJK test pins that trade-off.
Entry-level first-fit was declined because the group sentence would need two code-owned variants.
Findings-first folding was declined because it needs an unmeasured minimum-prefix constant and flips priority at the trigger.
A safety margin below the platform maximum was declined because no measurement supports a number.

## Tests

`TestSummaryBodyBudget` hand-types every limit, label, count, byte size and expected body; no fixture size or expected value comes from the code under test.
It pins the exact-fit and one-byte-over boundaries for both posters, and an exact-fit and one-byte-misfit pair binds every reservation term.
It also pins digit widths, list-order priority against best-fit, largest-first and smallest-first packing, and whole-group fitting through the real posters.
Fold tests cover line boundaries, code points, fences, repeated and overlapping HTML-comment openers, and fold priority.
Other tests cover bounded-path neutralization and forgery, the canonical footer, multi-byte guard refusals with exit 1, and stdout and stderr reporting.
Bench mirrors add one omission wiring case per platform, and the three frozen adapter snapshots stay byte-identical.
Mutation ledgers: build 29 of 29 red, fix round 1 19 of 20 red, fix round 2 14 of 14 red, which pinned the remaining overlapping-opener case, and fix round 3 5 of 5 red across the packing policies.

## Verification

- `node --test workflows/test/*.test.js` under Node 24.20.0: 1,380 pass, 0 fail.
- `node workflows/build.js && git diff --exit-code workflows/pipeline.js`: build fixed point ok.
- The generator is current, and Biome is ok.
- `pytest tests/`: 2,192 passed; coverage 93.99 percent against the 93.8 floor.
- `pytest bench/tests/`: 655 passed; coverage 89.55 percent against the 88.2 floor.
- `pre-commit run --all-files`: every hook passes.
- CI: 17 of 17 checks pass at `3ab5ecc`.

## Review record

A Codex GPT-6 Astra task did the recon.
Codex Astra and Claude opus red-teamed the design with mechanism and oracle lenses; all four asked for revision and none rejected.
Codex luna built the change and ran two fix rounds.
Each review round used Codex luna mutation and conformance lenses plus a Claude opus correctness lens.
Round 1 fuzzed 10,025 inputs against the base composition: no body over the limit, and every fitting input byte-identical.
Rounds 1 and 2 asked for fixes to test oracles, comments, docstrings and two doc paragraphs only.
Round 3 passed both Codex lenses with zero findings; the Claude lens found one packing fixture that no longer separated first-fit from best-fit, fixed in a third round that pins all five policies.
Lee's own code-gauntlet run is the cross-family PR pass.

## Adjacent issues

- #326 tracks per-comment body limits for GitHub inline review comments and GitLab discussions.
- #327 tracks off-enum severity labels that bypass the Summary display fold in the JS renderer.
- #328 tracks closing code fences by CommonMark rules in both prose folds.

## Not in this PR

- Inline comment and discussion body limits are in #326.
- Severity normalization is in #327.
- Four-backtick and tilde fence handling in the folds is in #328.
- No JS change is included.
- No paired mini is included; the Wave 5 checkpoint runs it.
- Bench snapshots are not regenerated.

Closes #324.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01WuLfVwNCvPBPW3fTWDoWw4

---

> [!NOTE]
> **Medium Risk**
> Changes core delivery composition and posting gates for GitHub/GitLab summary bodies; over-budget behavior can omit skipped findings or fold summary prose, though extensive tests and refusal-before-POST reduce silent API failures.
> 
> **Overview**
> **`post_review.py` now enforces per-platform UTF-8 byte limits** on the full GitHub review body and GitLab summary note (65,536 / 1,000,000). Bodies that already fit are unchanged; when over budget, the composer reserves header/footer/frame, **folds** oversized `review_body` (line-aware, UTF-8-safe), and admits **skipped consolidation groups** whole in list order—or omits them with a single “not shown” line—then always appends the **canonical** marker footer on the bounded path. Posters call **`_refuse_over_limit`** before the summary POST and print budget warnings on stdout/stderr.
> 
> **API shift:** `compose_review_body` takes `skipped_groups` plus platform metadata and returns **`ComposedBody`** (shown/omitted/fold counts); GitHub/GitLab paths build **grouped** skipped entries instead of flat lists. Bench payload mirrors use the same composer; **`TestSummaryBodyBudget`** and two mirror wiring tests lock behavior. Skill/docs now describe the budget instead of an uncapped skipped section.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 3ab5ecc4359c6bb3b679604309cff405267181b5. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by leehopper** [comment]
@liatrio-software-factory Approve

**Review by leehopper — COMMENTED** [review]
### ⚔️ Code Gauntlet

This PR (1570 authoritative changed lines) claims to introduce a per-platform UTF-8 byte budget for the complete posted PR/MR summary comment in scripts/post_review.py, replacing the previously "uncapped" skipped-findings section with a bounded composition scheme. GitHub is capped at 65536 bytes and GitLab at 1,000,000 bytes; compose_review_body now reserves the brand header, skipped-section frame, closing-count line, and canonical footer, then either posts the body whole (fast path, byte-identical to prior behavior) or falls into a bounded path that first tries to fold review_body prose (code-point-safe, fence-aware, HTML-comment-aware) and, when the summary still doesn't fit, admits whole consolidation groups of skipped findings in first-fit list order while omitting the rest with a summary closing line — footer placement and forgery-defence dedup semantics differ between the fast path (dedups against original review_body) and bounded path (always appends the full footer, with skipped-entry text excluded from the dedup). A new `_refuse_over_limit` guard makes the poster die before `post_json` if the composed body still exceeds budget after composition, and stdout/stderr reporting is extended (`_report_summary_budget`) to surface omitted-entry counts and folded-byte counts per platform. The stated motivation is closing a gap where a sufficiently large review body or skipped-findings section could silently exceed GitHub's/GitLab's real API body-size limits, which the prior uncapped design did not guard against. Risk is concentrated in scripts/post_review.py (flagged high risk), which is exercised by an extensive new oracle test class (TestSummaryBodyBudget) in tests/test_post_review.py covering fast-path/bounded-path boundaries, first-fit group admission policy, UTF-8 vs code-point measurement, fence/HTML-comment folding edge cases, forged-marker neutralization timing, and footer-dedup/refusal behavior at both platform limits; bench/tests/test_adapter.py's reference-payload mirrors were updated in parallel to call the new compose_review_body signature (skipped_groups instead of a flat skipped_entries list, with platform/findings_count/sha/inline_count keyword arguments) so the benchmark harness's independent payload reconstruction stays byte-identical to the real posters, adding two new budget-omission parity tests. Four reference docs (delivery-guide.md, headless-mode.md, phase8-delivery.md, report-format.md) were updated to replace prior claims that the skipped-finding section and full posted comment are "uncapped" with prose describing the new reservation order (header/footer first, summary folds only if needed, skipped groups shown whole-or-not-at-all in list order, one closing line, footer always last) — these are documentation-only changes describing the mechanism now implemented in post_review.py rather than code changes themselves.

3 finding(s) after the gauntlet — 3 low. 3 routed as improvement suggestion(s).

---
Generated by code-gauntlet | Reviewed up to: 3ab5ecc4359c6bb3b679604309cff405267181b5

File: scripts/post_review.py [hunk 1]
```diff
@@ -2038,11 +2237,20 @@ def post_github(data, valid_lines, line_texts):
...
+                f"  {composed.shown} of {len(flat_skipped)} finding(s) skipped inline "
+                "(lines not in diff) — appended to review body."
+            )
+        else:
+            skipped_line = (
+                f"  {len(flat_skipped)} finding(s) skipped inline (lines not in diff) — "
+                "appended to review body."
+            )
+        print(skipped_line)
+    _report_summary_budget(composed, "github")
```
Comment (leehopper) [inline scripts/post_review.py:2252]: **💡 [LOW] GitHub 'X of Y finding(s) skipped inline' stdout line untested when both shown and omitted are nonzero**

The new branch that prints composed.shown alongside len(flat_skipped) when composed.omitted is truthy is only exercised by a test where composed.shown happens to be 0, so a bug that always renders 0 (or swaps shown/omitted) would not be caught.

**Suggested fix:**
Add a poster-level test (via _run_poster or an equivalent full post_github dry-run) with one skipped group that fits in the budget and another that does not, then assert stdout contains the exact 'N of M finding(s) skipped inline' line with N>0 and N<M, e.g. building on test_consolidation_group_is_the_fitting_unit_through_main's fixture but also asserting the inline-skip stdout line.

⚔️ *Code Gauntlet*

File: skills/code-gauntlet/references/delivery-guide.md [hunk 2]
```diff
@@ -190,7 +192,7 @@ python3 {plugin_root}/scripts/post_review.py \
 - **GitLab:** Posts a summary note, then per-finding inline discussions with position metadata.
 - **Diff validation:** Parses diff to verify each finding line is in the PR/MR. Skips invalid lines with warning.
 - **Exit status:** Non-zero when nothing could be delivered, and — GitLab only — when a `--dry-run` finds any malformed inline position. The dry-run payload is still written; read it for what was wrong.
```
Comment (leehopper) [inline skills/code-gauntlet/references/delivery-guide.md]: **💡 [LOW] Exit-status bullet not updated for the new over-budget refusal path**

This PR adds a brand-new hard-failure path (_refuse_over_limit -> die(), exit code 1, nothing posted) that fires on both GitHub and GitLab whenever the composed body exceeds its platform byte budget, and unlike the one exit-status caveat this doc already calls out by name, the dry-run payload is NOT written in that case (post_json, which captures it, never runs). The Exit status bullet still only enumerates "nothing could be delivered" and the GitLab-only malformed-position case, so a reader following this doc has no warning that an over-budget dry-run leaves no payload to inspect.

**Suggested fix:**
Add a clause noting that an over-budget composed body also exits non-zero on both platforms and that, unlike the GitLab malformed-position case, no dry-run payload is written for that failure.

⚔️ *Code Gauntlet*

File: scripts/post_review.py [hunk 3]
```diff
@@ -1706,94 +1727,267 @@ def render_group_body(primary, corroborators, *, fence_offsets=None):
...
+
+    dropped_bytes = total - _utf8_len(prefix)
+    if prefix.count("```") % 2:
+        prefix += "" if prefix.endswith("\n") else "\n"
+        prefix += "```"
+    fold_line = (
+        f"_[folded: {dropped_bytes} more bytes; this {limits['surface']} reached the "
+        f"{limits['bytes']}-byte {limits['label']} body limit]_"
+    )
+    return f"{prefix}\n\n{fold_line}", dropped_bytes
```
Comment (leehopper) [inline scripts/post_review.py]: **💡 [LOW] Duplicate fold-line format string computed twice in _fold_review_body**

reserved_fold_line and fold_line build the exact same f-string template, once with total and once with dropped_bytes, so the wording lives in two places that must be kept byte-identical by hand; a local helper computing the line from a byte count removes the duplication with no behavior change.

**Suggested fix:**
Before: reserved_fold_line = f"_[folded: {total} more bytes; ...]_" ... fold_line = f"_[folded: {dropped_bytes} more bytes; ...]_" (same template twice). After: def fold_line_for(byte_count): return f"_[folded: {byte_count} more bytes; this {limits['surface']} reached the {limits['bytes']}-byte {limits['label']} body limit]_"; then use fold_line_for(total) and fold_line_for(dropped_bytes).

````suggestion
def _fold_review_body(text, allowance, platform):
    """Fold *text* into *allowance* bytes while preserving lines and code points."""
    limits = _body_limit(platform)
    total = _utf8_len(text)

    def fold_line_for(byte_count):
        return (
            f"_[folded: {byte_count} more bytes; this {limits['surface']} reached the "
            f"{limits['bytes']}-byte {limits['label']} body limit]_"
        )

    reserve = _utf8_len(f"\n\n{fold_line_for(total)}") + 4
    if allowance < reserve:
        prefix = ""
    else:
        prefix_allowance = allowance - reserve
        prefix = ""
        for index, line in enumerate(text.split("\n")):
            next_part = line if index == 0 else f"\n{line}"
            remaining = prefix_allowance - _utf8_len(prefix)
            if _utf8_len(next_part) <= remaining:
                prefix += next_part
                continue
            if _utf8_len(line) > prefix_allowance:
                prefix += _codepoint_prefix(next_part, remaining)
            break

    position = 0
    while True:
        opener = prefix.find("<!--", position)
        if opener < 0:
            break
        closer = prefix.find("-->", opener + 4)
        if closer < 0:
            prefix = prefix[:opener]
            break
        position = closer + 3

    dropped_bytes = total - _utf8_len(prefix)
    if prefix.count("```") % 2:
        prefix += "" if prefix.endswith("\n") else "\n"
        prefix += "```"
    return f"{prefix}\n\n{fold_line_for(dropped_bytes)}", dropped_bytes
````

⚔️ *Code Gauntlet*

File: scripts/post_review.py (hunk 1 above)
Comment (leehopper) [reply scripts/post_review.py:2252]: Fixed in 03683b4. `test_consolidation_group_is_the_fitting_unit_through_main` now asserts the exact line `  1 of 3 finding(s) skipped inline (lines not in diff) — appended to review body.`, so the branch is pinned with N>0 and N<M. Mutation-verified: rendering `composed.omitted` instead of `composed.shown` makes it print `2 of 3` and turns the test red.

File: skills/code-gauntlet/references/delivery-guide.md (hunk 2 above)
Comment (leehopper) [reply skills/code-gauntlet/references/delivery-guide.md]: Fixed in 03683b4. The exit-status bullet now says the composed body being over the platform byte budget exits non-zero on both platforms, and that the refusal happens before the post so no dry-run payload is written for it.

File: scripts/post_review.py (hunk 3 above)
Comment (leehopper) [reply scripts/post_review.py]: Fixed in 03683b4, as suggested: `_fold_review_body` now builds both the reserved and the emitted fold line from a single local `fold_line_for(byte_count)` helper. No behavior change; the fold tests stay green.
